consumer-services/src/main/scala/za/co/absa/spline/consumer/service/internal/AttributeDependencySolver.scala

```diff
@@ -19,72 +19,69 @@ package za.co.absa.spline.consumer.service.internal
 
-  def resolveDependencies(operations: Seq[OperationWithSchema], attributeID: UUID): AttributeDependencies = {
+  def resolveDependencies(operations: Seq[OperationWithSchema], attributeID: UUID): AttributeGraph = {
 
     val operationsMap = operations.map(op => op._id -> op).toMap
     val inputSchemaResolver = createInputSchemaResolver(operations)
 
-    def findDependenciesStart(
+    def findDependencies(
       operation: OperationWithSchema,
-      attributeID: UUID
-    ): (Set[UUID], Set[String]) = {
-
-      if (operation.schema.contains(attributeID))
-        findTransitiveDependencies(operation, Set(attributeID))
-      else if (operation.childIds.isEmpty)
-        (Set.empty, Set.empty)
-      else
-        operation
-          .childIds
-          .map(operationsMap)
-          .map(findDependenciesStart(_, attributeID))
-          .reduce(elementwiseUnion(_, _))
-    }
+      lookingFor: Set[UUID]
+    ): (Set[(UUID, UUID)], Map[UUID, String]) = {
 
-    def findTransitiveDependencies(
-      operation: OperationWithSchema,
-      attributeDependencies: Set[UUID]
-    ): (Set[UUID], Set[String]) = {
+      val (found, stillLookingFor) = {
+        val inputSchemas = inputSchemaResolver(operation)
+        val outputSchema = operation.schema
+
+        lookingFor.partition(id => outputSchema.contains(id) && !inputSchemas.exists(_.contains(id)))
+      }
+
+      val newNodes = found.map(_ -> operation._id).toMap
 
       val dependencyMap = resolveDependencies(operation, inputSchemaResolver)
-      val updatedAttributeDependencies = attributeDependencies
-        .flatMap(dependencyMap.get(_))
-        .flatten
-        .union(attributeDependencies)
+      val newEdges = found
+        .flatMap(x => dependencyMap.get(x).map(x -> _))
+        .flatMap { case (k, v) => v.map(inv => (k, inv)) }
+
+      val newLookingFor = stillLookingFor union newEdges.map(_._2)
 
-      val (attrDep, opDep) = operation
+      val (childEdges, childNodes) : (Set[(UUID, UUID)], Map[UUID, String]) = operation
         .childIds
         .map(operationsMap)
-        .filter(_.schema.exists(updatedAttributeDependencies))
-        .map(findTransitiveDependencies(_, updatedAttributeDependencies))
-        .reduceOption(elementwiseUnion(_, _))
-        .getOrElse((updatedAttributeDependencies, Set.empty[String]))
+        .filter(_.schema.exists(newLookingFor))
+        .map(findDependencies(_, newLookingFor))
+        .reduceOption(mergeResults(_, _))
+        .getOrElse((Set.empty, Map.empty))
 
-      (attrDep, opDep + operation._id)
+      (childEdges union newEdges, childNodes ++ newNodes)
     }
 
-    def elementwiseUnion[A, B](x: (Set[A], Set[B]), y: (Set[A], Set[B])): (Set[A], Set[B]) =
-      (x._1 union y._1, x._2 union y._2)
+    def mergeResults[A, B, C](x: (Set[A], Map[B,C]), y: (Set[A], Map[B,C])): (Set[A], Map[B,C]) =
+      (x._1 union y._1, x._2 ++ y._2)
+
+    val startOperation = operations.find(_.schema.contains(attributeID)).get
```

Comment:
Using `get` or `head` on monads is strongly discouraged. The reason is that it puts an assumption on a particular state of the monad, but tells you nothing about how to handle the situation when this assumption isn't met.
Prefer staying with a monad as long as possible and only unwrap them in the very end of your function or the entire logical unit.
Also, prefer `getOrElse( sys.error(...) )` or `getOrElse(throw ...)` to just `get` as it often gives the reader more context and makes troubleshooting easier if things go wrong.